onnxruntime/core/mlas/lib/amd64/cvtfp16Avx2.asm

```diff
@@ -0,0 +1,143 @@
+;++
+;
+; Copyright (c) Intel Corporation. All rights reserved.
+;
+; Licensed under the MIT License.
+;
+; Module Name:
+;
+;   cvtfp16Avx2.asm
+;
+; Abstract:
+;
+;   This module implements routines to convert between FP16 and FP32 formats using the AVX_NE_CONVERT ISA.
+;
+;--
+
+        .xlist
+INCLUDE mlasi.inc
+        .list
+
+        SUBTTL  "Convert buffer of half-precision floats to single-precision floats"
+;++
+;
+; Routine Description:
+;
+;   This routine converts the source buffer of half-precision floats to the
+;   destination buffer of single-precision floats.
+;
+;   This implementation uses AVX2 instructions.
+;
+; Arguments:
+;
+;   Source (rcx) - Supplies the address of the source buffer of half-precision
+;       floats.
+;
+;   Destination (rdx) - Supplies the address of the destination buffer of
+;       single-precision floats.
+;
+;   Count (r8) - Supplies the number of elements to convert.
+;
+; Return Value:
+;
+;   None.
+;
+;--
+
+        LEAF_ENTRY MlasConvertHalfToFloatBufferAVX2, _TEXT
+
+        test    r8, r8
+        jz      ExitRoutine
+        cmp     r8, 8
+        jb      ConvertMaskedVectors
+        cmp     r8, 16
+        jb      Convert128Vectors
+
+
+
+Convert256Vectors:
+        vcvtneeph2ps    ymm0, ymmword PTR [rcx]         ; Load even indexes
+        vcvtneoph2ps    ymm1, ymmword PTR [rcx]         ; Load odd indexes
+        vunpcklps       ymm2, ymm0, ymm1                ; Interleave low part
+        vunpckhps       ymm1, ymm0, ymm1                ; Interleave high part
+        vperm2f128      ymm0, ymm2, ymm1, 00100000b   	; Fix the order 
+        vperm2f128      ymm1, ymm2, ymm1, 00110001b   	; Fix the order 
+        vmovups         ymmword PTR [rdx], ymm0         ; Store the low part 
+        vmovups         ymmword PTR [rdx + 8*4], ymm1   ; Store the high part 
+
+        add     rcx, 16*2   ; Advance src ptr by 16 elements
+        add     rdx, 16*4   ; Advance dest ptr by 16 elements
+        sub     r8, 16      ; Reduce the counter by 16 elements
+
+        jz      ExitRoutine ; If we are done, exit
+        cmp     r8, 16      ; If the vector is big enough, we go again          
+        jae     Convert256Vectors   
+
```

Comment:
This instantly jumps to the Convert128Vectors do you need to check if there is less than 8 items left and jump to ConvertMaskedVectors in that case?  (What if you have less than 8 items left.)

Reply:
Because of how the instruction works, if there are less than 8 elements we still need to read and convert 8 because of the register size, but then we choose how many we need to store, that's what the Convert upper,l ower tags and the mask do

Reply:
You are right, will fix it